src/platforms/linux/daemon/dbusservice.cpp

```diff
@@ -130,11 +136,162 @@ QString DBusService::getLogs() {
+  QString portString = QString::number(port);
+
+  struct addrinfo hints;
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = AF_UNSPEC;
+  hints.ai_socktype = SOCK_DGRAM;
+  hints.ai_protocol = IPPROTO_UDP;
+
+  struct addrinfo* resolved = nullptr;
+  int retries = 15;
+
+  for (unsigned int timeout = 1000000;;
+       timeout = std::min((unsigned int)20000000, timeout * 6 / 5)) {
+    int rv = getaddrinfo(address.toLocal8Bit(), portString.toLocal8Bit(),
+                         &hints, &resolved);
+    if (!rv) {
+      break;
+    }
+
+    /* The set of return codes that are "permanent failures". All other
+     * possibilities are potentially transient.
+     *
+     * This is according to https://sourceware.org/glibc/wiki/NameResolver which
+     * states: "From the perspective of the application that calls getaddrinfo()
+     * it perhaps doesn't matter that much since EAI_FAIL, EAI_NONAME and
+     * EAI_NODATA are all permanent failure codes and the causes are all
+     * permanent failures in the sense that there is no point in retrying
+     * later."
+     *
+     * So this is what we do, except FreeBSD removed EAI_NODATA some time ago,
+     * so that's conditional.
+     */
+    if (rv == EAI_NONAME || rv == EAI_FAIL ||
+#ifdef EAI_NODATA
+        rv == EAI_NODATA ||
+#endif
+        (retries >= 0 && !retries--)) {
+      logger.log() << "Failed to resolve the address endpoint";
+      return false;
+    }
+
+    logger.log() << "Trying again in" << (timeout / 1000000.0) << "seconds";
+    usleep(timeout);
+  }
+
+  if ((resolved->ai_family == AF_INET &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in)) ||
+      (resolved->ai_family == AF_INET6 &&
+       resolved->ai_addrlen == sizeof(struct sockaddr_in6))) {
+    memcpy(endpoint, resolved->ai_addr, resolved->ai_addrlen);
+    freeaddrinfo(resolved);
+    return true;
+  }
+
+  logger.log() << "Invalid endpoint" << address;
+  freeaddrinfo(resolved);
+  return false;
+}
+
+bool DBusService::switchServer(const Config& config) {
+  logger.log() << "Switching server";
+
+  wg_device* device = static_cast<wg_device*>(calloc(1, sizeof(*device)));
+  if (!device) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  auto guard = qScopeGuard([&] { wg_free_device(device); });
+
+  strncpy(device->name, WG_INTERFACE, IFNAMSIZ - 1);
+  device->name[IFNAMSIZ - 1] = '\0';
+
+  wg_peer* peer = static_cast<wg_peer*>(calloc(1, sizeof(*peer)));
+  if (!peer) {
+    logger.log() << "Allocation failure";
+    return false;
+  }
+
+  peer->flags =
+      (wg_peer_flags)(WGPEER_HAS_PUBLIC_KEY | WGPEER_REPLACE_ALLOWEDIPS);
+
+  device->first_peer = device->last_peer = peer;
+  device->flags = WGDEVICE_REPLACE_PEERS;
+
+  if (!parseEndpoint(&peer->endpoint.addr, config.m_serverIpv4AddrIn,
+                     config.m_serverPort)) {
+    return false;
+  }
+
+  wg_key_from_base64(peer->public_key, config.m_serverPublicKey.toLocal8Bit());
```

Comment:
it's probably just me but feels like a loverhead to just know that the public key needs to be converted `toLocal8Bit` in order to be read `from_base64` can we make this more obvious somehow? or is this a reasonable thing to expect people to know. sorry for not knowing, just trying to calibrate.

Reply:
This is a good point. In reality, I call `toLocal8bit()` because it's the simplest way to obtain a `const char*` from a QString. Another approach is to have a QByteArray and use `constData()`.
I'll add a comment here too.